### ISSUE TITLE
Training is Slow

### ISSUE THREAD

**Issue by mmaaz60** [opencv/openvino_training_extensions#253]
Dear Team,

I am training MobileNet SSD model using NNCF. It looks like the framework does not utilize the hardware enough and hence the training is really slow.

Note below my system specs and training configurations.
* GPU: RTX 2080 Supper (8 GB)
* Batch Size = 32
* Configuration File: [click here](https://github.com/opencv/openvino_training_extensions/blob/develop/pytorch_toolkit/nncf/examples/object_detection/configs/ssd_mobilenet_voc.json)
* Dataset: VOC 2007 + 2012 as described [here](https://github.com/opencv/openvino_training_extensions/tree/develop/pytorch_toolkit/nncf/examples/object_detection)
* Average Epoch Time: 15 minutes

I have recorded a [video](https://drive.google.com/file/d/1FwAAaXR2GHp0P-fAU4zn7l4UGQbi8MNO/view?usp=sharing) capturing the GPU utilization while training. Note that the GPU utilization remains below 10%. I have been using [AlexeAB/darknet](https://github.com/mmaaz60/darknet) clone for the training of yolov3 variants. It is quite fast.

Please guide me if I am missing anything.

**Stay Safe,**
Muhammad Maaz

**Comment by AlexanderDokuchaev** [comment]
@AlexKoff88 @vshampor @ljaljushkin

**Comment by AlexKoff88** [comment]
Hi,

Do you use any optimization/compression method during your experiments? or you just trained a full-precision model?

Can you please share the training config, logs, and so on?

Regards,
Alexander

**Comment by mmaaz60** [comment]
> Hi,
> 
> Do you use any optimization/compression method during your experiments? or you just trained a full-precision model?

Thank You @AlexKoff88,

For this, I am not using any compression method and training a full-precision model.

> Can you please share the training config, logs, and so on?

The training logs are [ssd_mobilenet_output.log](https://github.com/opencv/openvino_training_extensions/files/4367799/ssd_mobilenet_output.log). Note that configuration parameters are all written in the training logs.


I am also training the same model with rb_sparsity and quantization (in8) compression and it is taking around 3x time than the full precision model. The training log file is [ssd_mobilenet_sparsity_int8_output.log](https://github.com/opencv/openvino_training_extensions/files/4367826/ssd_mobilenet_sparsity_int8_output.log).

Please let me know if I am missing something.

**Stay Safe,**
Maaz

**Comment by vshampor** [comment]
Greetings, @mmaaz60!

I checked the configuration you provided on an RTX 2080 Ti 12 GB. The utilization of GPU rises as high as 30% during training, see the GIF attached:

![mov](https://user-images.githubusercontent.com/31695470/77298927-b73ce080-6cfc-11ea-88f3-c31d995b698a.gif)

Epoch iteration runtimes as reported in logs are:
```
0: iter 0 epoch 0 || Loss: 18.57 || Time 1.076s || lr: 0.0001 || CR loss: 0
0: iter 10 epoch 0 || Loss: 14.57 || Time 4.502s || lr: 0.0001 || CR loss: 0
0: iter 20 epoch 0 || Loss: 13.17 || Time 4.367s || lr: 0.0001 || CR loss: 0
0: iter 30 epoch 0 || Loss: 11.44 || Time 9.886s || lr: 0.0001 || CR loss: 0
0: iter 40 epoch 0 || Loss: 10.5 || Time 10.87s || lr: 0.0001 || CR loss: 0
0: iter 50 epoch 0 || Loss: 9.293 || Time 9.842s || lr: 0.0001 || CR loss: 0
0: iter 60 epoch 0 || Loss: 9.022 || Time 10.03s || lr: 0.0001 || CR loss: 0
0: iter 70 epoch 0 || Loss: 10.04 || Time 9.695s || lr: 0.0001 || CR loss: 0
0: iter 80 epoch 0 || Loss: 9.031 || Time 8.049s || lr: 0.0001 || CR loss: 0
0: iter 90 epoch 0 || Loss: 9.305 || Time 13.18s || lr: 0.0001 || CR loss: 0
0: iter 100 epoch 0 || Loss: 8.867 || Time 10.11s || lr: 0.0001 || CR loss: 0
0: iter 110 epoch 0 || Loss: 8.776 || Time 9.504s || lr: 0.0001 || CR loss: 0
0: iter 120 epoch 0 || Loss: 8.272 || Time 7.771s || lr: 0.0001 || CR loss: 0
0: iter 130 epoch 0 || Loss: 8.037 || Time 9.898s || lr: 0.0001 || CR loss: 0
0: iter 140 epoch 0 || Loss: 8.846 || Time 13.48s || lr: 0.0001 || CR loss: 0
0: iter 150 epoch 0 || Loss: 8.812 || Time 8.932s || lr: 0.0001 || CR loss: 0
0: iter 160 epoch 0 || Loss: 8.325 || Time 8.103s || lr: 0.0001 || CR loss: 0
```

I also observe a 3x increase in training time while enabling INT8 and RB sparsity compression.

Our HW is slightly different - what I might suggest to you is rolling back to the CUDA toolkit version 10.1 (driver version is visible in the GIF). That the model training does not utilize GPU completely might be normal if the computations are memory/disk bound rather than compute-bound - you might want to move your dataset to an SSD instead of HDD. 

As for performance comparisons with YOLOv3, I don't think that these are valid arguments for the point that MobileNet-SSD training in this repo is slow. Please provide correct performance measurements comparing training performance in NNCF repo and in another, reference repo while using exactly the same model architecture and HW. If NNCF object detection performance is really worse in that case, we will look into it.

**Comment by ljaljushkin** [comment]
@vshampor have you trained with DataParallel or DistributedDataParallel?
I'd suggest trying the latter mode with more workers by specifying: 
`
--multiprocessing-distributed
--workers 12
`

**Comment by vshampor** [comment]
@mmaaz60, after taking a look at the object detection training example script I found a bug that might impact training time - the PyTorch checkpoints were saved each iteration instead of each epoch. The fix can be found at: https://github.com/opencv/openvino_training_extensions/pull/255
Please check whether it speeds up the training process for you in any appreciable manner.

**Comment by mmaaz60** [comment]
> @mmaaz60, after taking a look at the object detection training example script I found a bug that might impact training time - the PyTorch checkpoints were saved each iteration instead of each epoch. The fix can be found at: #255
> Please check whether it speeds up the training process for you in any appreciable manner.

Hi @vshampor,

It works. For **SSD_MobileNet with RB Sparsity + Quantization (INT8)** compression aware training, I got around **18x speed gain**. GPU utilization is now between 50-60%. I will also look at the code and try to tune it for speed and share my findings if any. Thank you

**Stay Safe,**
Maaz
